api/src/main/java/jakarta/resource/spi/work/WorkContextErrorCodes.java

```diff
@@ -29,6 +29,12 @@
  */
 public class WorkContextErrorCodes {
 
+	/**
+	 * Default constructor.
+	 */
+	protected WorkContextErrorCodes() {
+	}
```

Comment:
The change to `protected` is a breaking change to any code that currently does `new WorkContextErrorCodes()`.  Although it does seem unlikely that code would ever bother instantiating this class, if there is no benefit to be gained from the switch to `protected`, we should avoid taking the risk.

Reply:
right but still [within constraints](https://eclipse-ee4j.github.io/jakartaee-platform/CompatibilityRequirements). The benefit is `forcing` users to access static fields properly and not through an instance object

Reply:
this has no impact on applications, only on vendors where _it is expected that a new version of the specification will require existing vendor implementations to be modified and extended to implement the new version_

Or is this SPI to be used by applications directly?

Reply:
I agree we don't need to be concerned with application usage (2) because Work is an SPI between the vendor spec implementation (1) - which we don't need to maintain compatibility for - and the resource adapter vendor (3) - which we do need to maintain compatibility for, albeit with some amount of gray area due to the "less absolute" language.  However, when I read question Q4, it seems to clarify a violation of (3) for this.  I'd certainly like to say we can just change it because we don't think code would be following this discouraged pattern of an instantiating the class to access constants from it, but I didn't see allowances for that in the document.  If others want to overrule me on this, that's fine. I just wanted to be sure to point it out.

Reply:
TBH, there are specs within EE 10 (even EE 9.1) which are removing even application-level APIs (through the major version change). In a strict world, even increase of the binary level should come with the major version change I think

Reply:
Yes, breaking changes are fine if updating the major version, so another option here would be to deprecate these unwanted public constructors in 2.1 with the plan to make them private in 3.0.

Reply:
JPMS requires  - and there is compile time warning currently for this in SE 17+ - some explicitly defined public or protected constructor for exported classes (the compiler generated one is not enough), so deprecating & making them private is not an option.

Reply:
see https://bugs.openjdk.java.net/browse/JDK-8071961

Reply:
I'm going to revert this part of the change. We're wasting time here for nothing...